libsolidity/ast/Types.cpp

```diff
@@ -199,18 +148,78 @@ util::Result<TypePointers> transformParametersToExternal(TypePointers const& _pa
 	return transformed;
 }
 
+bool compareTypes(Type const* lhs, Type const* rhs)
+{
+	solAssert(lhs && rhs, "");
+	return lhs->richIdentifier() < rhs->richIdentifier();
 }
+using TypeSet = std::set<Type const*, decltype(compareTypes)*> ;
 
-vector<frontend::Type const*> solidity::frontend::oversizedSubtypes(frontend::Type const& _type)
+void storageOversizedSubtypesInner(
+	Type const* _type,
+	bool _includeType,
+	set<string>& _seen,
+	TypeSet& _oversizedSubtypes
+)
 {
-	set<StructDefinition const*> structsSeen;
-	TypeSet oversized;
-	oversizedSubtypesInner(_type, false, structsSeen, oversized);
-	vector<frontend::Type const*> res;
+	for (Type const* subtype: _type->subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (_seen.count(id))
+			continue;
+		_seen.insert(id);
+		bool includeType = _type->category() != Type::Category::Struct;
+		storageOversizedSubtypesInner(subtype, includeType, _seen, _oversizedSubtypes);
+		_seen.erase(id);
+	}
+
+	if (_includeType && _type->storageSizeUpperBound() >= bigint(1) << 64)
+		_oversizedSubtypes.insert(_type);
+}
+
+}
+
+vector<Type const*> Type::storageOversizedSubtypes() const
+{
+	set<string> seen;
+	TypeSet oversized(compareTypes);
+	storageOversizedSubtypesInner(this, false, seen, oversized);
+	vector<Type const*> res;
 	copy(oversized.cbegin(), oversized.cend(), back_inserter(res));
 	return res;
 }
 
+BoolResult Type::validForLocation(DataLocation _loc, set<string>* _seen) const
+{
+	if (_seen == nullptr)
+	{
+		set<string> seen;
+		return validForLocation(_loc, &seen);
+	}
+
+	for (Type const* subtype: subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (_seen->count(id))
+			continue;
+		_seen->insert(id);
+		BoolResult result = subtype->validForLocation(_loc, _seen);
+		_seen->erase(id);
+		if (!result)
+			return result;
+	}
+
+	bigint size = sizeUpperBound(_loc);
+	if (_loc == DataLocation::Storage && size >= bigint(1) << 256)
+		return BoolResult::err("Type too large for storage.");
+	if (_loc == DataLocation::Memory && size >= numeric_limits<unsigned>::max())
+		return BoolResult::err("Type too large for memory.");
+	if (_loc == DataLocation::CallData && size >= numeric_limits<unsigned>::max())
+		return BoolResult::err("Type too large for calldata.");
```

Comment:
Wouldn't this be dependent on the size of `unsigned` on the system? Why don't we write `2**64` explicitly?

Reply:
Yes, probably better to use `2**64` here and `uint64_t` as the type where it is needed.

Reply:
`2**64` or `2**32`? Is not `unsigned` 32 bit typically?
On my system,  `numeric_limits<unsigned>::max() == 4294967295 == 2**32 - 1`.
So with `2**64` we might be relaxing current restraints.

Reply:
Right, I think it got conflated with the 64-bit check that it's the code - but that check is for storage, not memory and calldata. Anyway, I think that `unsigned` should be replaced with `uint32_t`. because we do need the size of that type to be independent of the underlying platform.

Reply:
Changed to `numeric_limits<uint32_t>::max()`.

Reply:
I understood @chriseth's comment to mean that not just this piece of code should use `uint32_t` but also e.g. the functions that return memory and calldata sizes in `Type`. Currently they mostly return `unsigned`. Changing it only here results in them being out of sync.

Actually this would add a lot of changes to this PR so I think it would be better to leave it here as `unsigned` after all (to keep the types in sync) and to have a separate PR that changes both this code and the functions

Reply:
OK, changed back to `unsigned`.